File: python/linpack/Linpack.py
import time, math

ARRAY_SIZE = 2000

def linpack(array_size : int):
  mflops_result = .0
  residn_result = .0
  time_result = .0
  eps_result = .0
  a = [[0] * array_size for i in range(array_size)]
  b = [0] * array_size
  x = [0] * array_size
  ipvt = [0] * array_size
  lda = array_size
  n = array_size

  ops = ((2.0e0 * n) * n * n) / 3.0 + 2.0 * (n * n)

  norma = __matgen(a, lda, n, b)

  start = time.time()

  info = __dgefa(a, lda, n, ipvt)
  __dgesl(a, lda, n, ipvt, b, 0)

  total = time.time() - start

  x = [i for i in b]

  norma = __matgen(a, lda, n, b)

  b = [-i for i in b]


  __dmxpy(n, b, n, lda, x, a)

  resid = .0
  normx = .0

  for i in range(n):
      resid = resid if resid > abs(b[i]) else abs(b[i])
      normx = normx if normx > abs(x[i]) else abs(x[i])

  eps_result = __epslon(1.0)
  residn_result = (resid / ((((n) * norma * normx) * eps_result)))
  residn_result += .005
  residn_result = (math.floor((residn_result * (100))))
  residn_result /= (100)
  time_result = total
  time_result += .005
  time_result = (math.floor((time_result * (100))))
  time_result /= (100)

  mflops_result = ops / (1.0e6 * total)
  mflops_result += .0005
  mflops_result = (math.floor((mflops_result * (1000))))
  mflops_result /= (1000)

  result = {
    "Norma" : norma,
    "Residual" : resid,
    "NormalisedResidual" : residn_result,
    "Eepsilon" : eps_result,
    "Time" : time_result,
    "MFLOPS" : mflops_result
  }

  print(result)

  return result

def __matgen(a, lda : int, n : int, b) -> float:
  iseed = [0] * 4
  iseed[0] = 1
  iseed[1] = 2
  iseed[2] = 3
  iseed[3] = 1325
  norma = .0

  for i in range(n):
    for j in range(n):
      a[j][i] = (__lran(iseed) - .5)
      norma = (a[j][i] if (a[j][i] > norma) else norma)

  for i in range(n):
    b[i] = 0.0

  for j in range(n):
    for i in range(n):
      b[i] += a[j][i]

  return norma

def __lran(seed) -> float:
  m1 = 494
  m2 = 322
  m3 = 2508
  m4 = 2549
  ipw2 = 4096
  r = 1.0 / ipw2
  it4 = seed[3] * m4
  it3 = math.floor(it4 / ipw2)
  it4 = it4 - ipw2 * it3
  it3 = it3 + seed[2] * m4 + seed[3] * m3
  it2 = math.floor(it3 / ipw2)
  it3 = it3 - ipw2 * it2
  it2 = it2 + seed[1] * m4 + seed[2] * m3 + seed[3] * m2
  it1 = math.floor(it2 / ipw2)
  it2 = it2 - ipw2 * it1
  it1 = it1 + seed[0] * m4 + seed[1] * m3 + seed[2] * m2 + seed[3] * m1
  it1 = it1 % ipw2
  seed[0] = it1
  seed[1] = it2
  seed[2] = it3
  seed[3] = it4
  result = r * (it1 + r * (it2 + r * (it3 + r * it4)))
  return result

def __dgefa(a, lda : int, n : int, ipvt) -> int:
  col_k = [ ]
  col_j = [ ]
  info = 0
  nm1 = (n - 1)
  if (nm1 >= 0): 
    for k in range(nm1):
      col_k = a[k]
      kp1 = (k + 1)
      l_ = (__idamax(n - k, col_k, k, 1) + k)
      ipvt[k] = l_
      if col_k[l_] != 0:
        if l_ != k:
          t = col_k[l_]
          col_k[l_] = col_k[k]
          col_k[k] = t
        t = -1.0 / col_k[k]
        __dscal(n - kp1, t, col_k, kp1, 1)
        for j in range(kp1, n):
          col_j = a[j]
          t = col_j[l_]
          if l_ != k: 
            col_j[l_] = col_j[k]
            col_j[k] = t
          __daxpy(n - kp1, t, col_k, kp1, 1, col_j, kp1, 1)
      else: 
        info = k
  ipvt[n - 1] = (n - 1)
  if (a[(n - 1)][(n - 1)] == 0): 
    info = (n - 1)
  return info

def __dgesl(a, lda : int, n : int, ipvt, b, job : int) -> None:
  nm1 = n - 1
  if job == 0: 
    if (nm1 >= 1): 
      for k in range(nm1):
        l_ = ipvt[k]
        t = b[l_]
        if l_ != k:
          b[l_] = b[k]
          b[k] = t
        kp1 = k + 1
        __daxpy(n - kp1, t, a[k], kp1, 1, b, kp1, 1)
    for kb in range(n):
      k = n - (kb + 1)
      b[k] /= a[k][k]
      t = - b[k]
      __daxpy(k, t, a[k], 0, 1, b, 0, 1)
  else: 
    for k in range(n): 
      t = __ddot(k, a[k], 0, 1, b, 0, 1)
      b[k] = (b[k] - t) / a[k][k]
    if (nm1 >= 1): 
      for kb in range(1, nm1):
        k = (n - ((kb + 1)))
        kp1 = (k + 1)
        b[k] += __ddot(n - ((kp1)), a[k], kp1, 1, b, kp1, 1)
        l_ = ipvt[k]
        if (l_ != k): 
          t = b[l_]
          b[l_] = b[k]
          b[k] = t

def __daxpy(n : int, da : float, dx, dx_off : int, incx : int, dy, dy_off : int, incy : int) -> None:
  if (((n > 0)) and ((da != 0))): 
    if (incx != 1 or incy != 1): 
      ix = 0
      iy = 0
      if (incx < 0): 
        ix = (- n + 1) * incx
      if (incy < 0): 
        iy = (- n + 1) * incy
      for i in range(n): 
        dy[iy + dy_off] += (da * dx[ix + dx_off])
        ix += incx
        iy += incy
      return
    else: 
      for i in range(n):
        dy[i + dy_off] += da * dx[i + dx_off]

def __ddot(n : int, dx, dx_off : int, incx : int, dy, dy_off : int, incy : int) -> float:
  dtemp = (0)
  if (n > 0): 
    if (incx != 1 or incy != 1): 
      ix = 0
      iy = 0
      if (incx < 0): 
        ix = (- n + 1) * incx
      if (incy < 0): 
        iy = (- n + 1) * incy
      for i in range(n):
        dtemp += dx[ix + dx_off] * dy[iy + dy_off]
        ix += incx
        iy += incy
    else: 
      for i in range(n):
        dtemp += dx[i + dx_off] * dy[i + dy_off]
  return (dtemp)

def __dscal(n : int, da : float, dx, dx_off : int, incx : int) -> None:
  if (n > 0): 
    if (incx != 1): 
      nincx = (n * incx)
      for i in range(0, nincx, incx):
        dx[i + dx_off] *= da
    else: 
      for i in range(n):
        dx[i + dx_off] *= da

def __idamax(n : int, dx, dx_off : int, incx : int) -> int:
  itemp = 0
  if (n < 1): 
    itemp = -1
  elif (n == 1): 
    itemp = 0
  elif (incx != 1): 
    dmax = abs(dx[0 + dx_off])
    ix = (1 + incx)
    for i in range(n):
      dtemp = abs(dx[ix + dx_off])
      if (dtemp > dmax): 
        itemp = i
        dmax = dtemp
      ix += incx
  else: 
    itemp = 0
    dmax = abs(dx[0 + dx_off])
    for i in range(n): 
      dtemp = abs(dx[i + dx_off])
      if (dtemp > dmax): 
        itemp = i
        dmax = dtemp
  return (itemp)

def __epslon(x : float) -> float:
  a = 4.0e0 / 3.0e0
  eps = (0)
  while eps == 0:
    b = (a - 1.0)
    c = 3*b
    eps = abs(c - 1.0)
  return (eps * abs(x))

def __dmxpy(n1 : int, y, n2 : int, ldm : int, x, m) -> None:
  for j in range(n2):
    i = 0
    for i in range(n1): 
      y[i] += (x[j] * m[j][i])
      i += 1

def run():
  linpack(ARRAY_SIZE)

run()

File: rust/linpack/Linpack.rs
use std::time::{Duration, Instant};

const ARRAY_SIZE: usize = 2000;

struct LinpackResult {
    norma: f64,
    residual: f64,
    normalised_residual: f64,
    epsilon: f64,
    time: f64,
    mflops: f64,
}

fn second() -> f64 {
    let now = Instant::now();
    now.elapsed().as_secs_f64()
}

use rand::Rng;

fn matgen(a: &mut Vec<Vec<f64>>, b: &mut Vec<f64>, n: usize) {
    assert_eq!(a.len(), n);
    assert_eq!(a[0].len(), n);
    assert_eq!(b.len(), n);

    let mut rng = rand::thread_rng();
    for i in 0..n {
        for j in 0..n {
            a[i][j] = rng.gen();
        }
        b[i] = rng.gen();
    }
}

fn lran(seed: &mut i32) -> f64 {
    const MULTIPLIER: i32 = 1366;
    const ADDEND: i32 = 150889;
    const PMOD: i32 = 714025;
    
    *seed = (*seed * MULTIPLIER + ADDEND) % PMOD;
    (*seed as f64) / PMOD as f64 - 0.5
}

fn dgefa(a: &mut Vec<Vec<f64>>, ipvt: &mut Vec<usize>, n: usize) {
    for k in 0..n-1 {
        let mut imax = k;
        let mut amax = a[k][k].abs();
        for i in k+1..n {
            let abs = a[i][k].abs();
            if abs > amax {
                imax = i;
                amax = abs;
            }
        }
        ipvt[k] = imax;
        if amax == 0.0 {
            panic!("Singular matrix");
        }
        if imax != k {
            a.swap(imax, k);
        }
        for i in k+1..n {
            a[i][k] /= a[k][k];
            for j in k+1..n {
                a[i][j] -= a[i][k] * a[k][j];
            }
        }
    }
    ipvt[n-1] = n-1;
    if a[n-1][n-1] == 0.0 {
        panic!("Singular matrix");
    }
}

fn dgesl(a: &Vec<Vec<f64>>, ipvt: &Vec<usize>, b: &mut Vec<f64>, n: usize) {
    for kb in 0..n {
        let k = n - (kb + 1);
        let l = ipvt[k];
        let t = b[l];
        b[l] = b[k];
        b[k] = t;
        for i in (k+1)..n {
            b[i] -= a[i][k] * t;
        }
    }
    for kb in 0..n {
        let k = n - (kb + 1);
        b[k] /= a[k][k];
        let t = -b[k];
        for i in 0..k {
            b[i] += a[i][k] * t;
        }
    }
}

fn daxpy(n: usize, da: f64, dx: &[f64], incx: usize, dy: &mut [f64], incy: usize) {
    let mut ix = 0;
    let mut iy = 0;
    for _ in 0..n {
        dy[iy] += da * dx[ix];
        ix += incx;
        iy += incy;
    }
}

fn ddot(n: usize, dx: &[f64], incx: usize, dy: &[f64], incy: usize) -> f64 {
    let mut ix = 0;
    let mut iy = 0;
    let mut dot = 0.0;
    for _ in 0..n {
        dot += dx[ix] * dy[iy];
        ix += incx;
        iy += incy;
    }
    dot
}

fn dscal(n: usize, da: f64, dx: &mut [f64], incx: usize) {
    let mut ix = 0;
    for _ in 0..n {
        dx[ix] *= da;
        ix += incx;
    }
}

fn idamax(n: usize, dx: &[f64], incx: usize) -> usize {
    let mut ix = 0;
    let mut imax = 0;
    let mut dmax = 0.0;
    for _ in 0..n {
        let abs = dx[ix].abs();
        if abs > dmax {
            imax = ix;
            dmax = abs;
        }
        ix += incx;
    }
    imax
}

fn linpack(array_size: usize) -> LinpackResult {
    let mut a = vec![vec![0.0; array_size]; array_size];
    let mut b = vec![0.0; array_size];
    let mut x = vec![0.0; array_size];
    let mut ipvt = vec![0; array_size];
    let mut norma = 0.0;
    let mut resid = 0.0;
    let mut residn = 0.0;
    let mut eps = 0.0;
    let mut time = 0.0;
    let mut mflops = 0.0;

    let mut seed = 1325;
    let mut kase = 0;
    let mut n = array_size;

    while n <= array_size {
        kase += 1;
        matgen(&mut a, &mut b, n);
        let start = second();
        dgefa(&mut a, &mut ipvt, n);
        dgesl(&a, &ipvt, &mut b, n);
        time += second() - start;
        for i in 0..n {
            x[i] = b[i];
        }
        norma = 0.0;
        for i in 0..n {
            for j in 0..n {
                norma = f64::max(norma, a[i][j].abs());
            }
        }
        for i in 0..n {
            b[i] = 0.0;
            for j in 0..n {
                b[i] += a[i][j] * x[j];
            }
        }
        resid = 0.0;
        for i in 0..n {
            resid = f64::max(resid, (x[i] - b[i]).abs());
        }
        eps = epslon(&1.0);
        residn = resid / (n as f64 * norma * eps);
        time = time / kase as f64;
        mflops = 2.0 * n as f64 * n as f64 * n as f64 / 3.0 / 1e6 / time;
        n += 1000;
    }

    LinpackResult {
        norma,
        residual: resid,
        normalised_residual: residn,
        epsilon: eps,
        time,
        mflops,
    }
}

fn epslon(x: &f64) -> f64 {
    let mut a = 1.0;
    let mut b = 1.0;
    while (a + 1.0) != 1.0 {
        b /= 2.0;
        a = b + 1.0;
    }
    b * 2.0
}

fn main() {
    let start_time = Instant::now();
    let result = linpack(ARRAY_SIZE);
    let duration = start_time.elapsed();

    println!("Linpack Result: {:?}", result);
    println!("Execution time: {:?}", duration);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_matgen() {
        let mut a = vec![vec![0.0; 3]; 3];
        let mut b = vec![0.0; 3];
        matgen(&mut a, &mut b, 3);
        assert_eq!(a.len(), 3);
        assert_eq!(a[0].len(), 3);
        assert_eq!(b.len(), 3);
    }

    #[test]
    fn test_lran() {
        let mut seed = 1325;
        let r = lran(&mut seed);
        assert!(r >= -0.5 && r < 0.5);
    }

    #[test]
    fn test_dgefa() {
        let mut a = vec![
            vec![1.0, 2.0, 3.0],
            vec![4.0, 5.0, 6.0],
            vec![7.0, 8.0, 9.0],
        ];
        let mut ipvt = vec![0; 3];
        dgefa(&mut a, &mut ipvt, 3);
        assert_eq!(ipvt, vec![2, 2, 2]);
    }

    #[test]
    fn test_dgesl() {
        let a = vec![
            vec![1.0, 2.0, 3.0],
            vec![4.0, 5.0, 6.0],
            vec![7.0, 8.0, 9.0],
        ];
        let ipvt = vec![2, 2, 2];
        let mut b = vec![1.0, 2.0, 3.0];
        dgesl(&a, &ipvt, &mut b, 3);
        assert_eq!(b, vec![-0.5, 1.0, 0.0]);
    }

    #[test]
    fn test_daxpy() {
        let mut dy = vec![1.0, 2.0, 3.0];
        let dx = vec![1.0, 2.0, 3.0];
        daxpy(3, 2.0, &dx, 1, &mut dy, 1);
        assert_eq!(dy, vec![3.0, 6.0, 9.0]);
    }

    #[test]
    fn test_ddot() {
        let dx = vec![1.0, 2.0, 3.0];
        let dy = vec![1.0, 2.0, 3.0];
        let dot = ddot(3, &dx, 1, &dy, 1);
        assert_eq!(dot, 14.0);
    }

    #[test]
    fn test_dscal() {
        let mut dx = vec![1.0, 2.0, 3.0];
        dscal(3, 2.0, &mut dx, 1);
        assert_eq!(dx, vec![2.0, 4.0, 6.0]);
    }

    #[test]
    fn test_idamax() {
        let dx = vec![1.0, 2.0, 3.0];
        let imax = idamax(3, &dx, 1);
        assert_eq!(imax, 2);
    }

    #[test]
    fn test_linpack() {
        let result = linpack(100);
        assert!(result.norma > 0.0);
        assert!(result.residual > 0.0);
        assert!(result.normalised_residual > 0.0);
        assert!(result.epsilon > 0.0);
        assert!(result.time > 0.0);
        assert!(result.mflops > 0.0);
    }
}
